### PR TITLE
feat: update to lib types for TS 5.2 RC

### PR DESCRIPTION
**Pull request by sosukesuzuki**
## PR Checklist

- [x] Addresses an existing open issue: RC support of #7155 
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

Update TypeScript to 5.2.1-rc

### CODE REVIEW

File: packages/typescript-estree/src/version-check.ts [hunk 1]
```diff
@@ -20,6 +20,8 @@ const versions = [
   '4.8',
   '4.9',
   '5.0',
+  '5.1',
+  '5.2',
```
Comment (JoshuaKGoldberg) [inline packages/typescript-estree/src/version-check.ts:24]: ...huh. Odd that 5.1 wasn't here before.

File: packages/scope-manager/tsconfig.spec.json [hunk 2]
```diff
@@ -2,8 +2,9 @@
   "extends": "./tsconfig.json",
   "compilerOptions": {
     "outDir": "../../dist/out-tsc",
-    "module": "commonjs",
-    "types": ["jest", "node"]
+    "module": "Node16",
+    "types": ["jest", "node"],
+    "moduleResolution": "Node16"
```
Comment (JoshuaKGoldberg) [inline packages/scope-manager/tsconfig.spec.json:7]: Hmm, why this change? Tests still work for me locally after reverting it.

**Review by JoshuaKGoldberg — DISMISSED** [review]
LGTM, thanks! I'll defer to @bradzacher as more knowledgeable on TS upgrades.

Note for any onlookers: we'll still have to add support for `using` statements separately.

**Review by bradzacher — DISMISSED** [review]
Thanks - step one is this! Step two is the new syntax.

File: packages/scope-manager/tsconfig.spec.json (hunk 2 above)
Comment (bradzacher) [reply packages/scope-manager/tsconfig.spec.json:7]: But does it typecheck fine without it?
There's a new error in this version to ensure that things are configured correctly.

Comment (JoshuaKGoldberg) [reply packages/scope-manager/tsconfig.spec.json:7]: Yeah I like moving to the newer module & module resolution settings, just curious.

Also, per https://www.typescriptlang.org/tsconfig#moduleResolution, `moduleResolution` matches `module` by default with `node16` and other newer targets.

Comment (sosukesuzuki) [reply packages/scope-manager/tsconfig.spec.json:7]: Without this change, a type error will occur.

Comment (JoshuaKGoldberg) [reply packages/scope-manager/tsconfig.spec.json:7]: Ah! https://devblogs.microsoft.com/typescript/announcing-typescript-5-2-rc/#module-and-moduleresolution-must-match-under-recent-node-js-settings

```plaintext
       tsconfig.spec.json:5:15 - error TS5110: Option 'module' must be set to 'Node16' when option 'moduleResolution' is set to 'Node16'.
       
       5     "module": "CommonJS",
                       ~~~~~~~~~~
```

Fun! In that case, I think we can get away with just the `"module": "Node16"`:

```suggestion
```

Comment (sosukesuzuki) [reply packages/scope-manager/tsconfig.spec.json]: Thank you! [e829631](https://github.com/typescript-eslint/typescript-eslint/pull/7451/commits/e82963179840354b8d091e452f2eff52dded63eb)

**Review by bradzacher — APPROVED** [review]
next we can get to work on the syntax thanks heaps!
